### ISSUE TITLE
oracledb not releasing connection from pool after disconnect

### ISSUE THREAD

**Issue by jakecoffman** [knex/knex#3605]
# Environment

Knex version: 0.20.6
Database + version: Oracle 12c
OS: Linux & OSX

@atiertant 

# Bug

1. Explain what kind of behaviour you are getting and how you think it should do:

When oracledb falls of the network or the application can't connect and there's an open transaction, the pool doesn't empty. After the pool is full we see `Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?` and the application cannot be used until we restart it. 

2. Error message

`Knex: Timeout acquiring a connection. The pool is probably full. Are you missing a .transacting(trx) call?`

3. Reduced test code, for example in https://npm.runkit.com/knex or if it needs real
   database connection to MySQL or PostgreSQL, then single file example which initializes
   needed data and demonstrates the problem.

See https://github.com/jakecoffman/knex-oracle-bug

The crux is: 

```
  const trx = await knex.transaction()
  console.log("Started transaction")

  await sleep(5000)
  console.log("Committing")

  try {
    await trx.commit()
  } catch (e) {
    console.log("Error committing:", e)
    await knex.destroy()
    return
  }

  console.log("Ok")
```

If you disconnect during the sleep then the `commit()` still finishes immediately (I expected it would block and throw on reject but it does not) and after a minute you see an Unhandled rejection and the connection isn't freed from the pool.

**Comment by elhigu** [comment]
Connection should not be removed from pool at that point yet. However it should be marked to be disposed next time when it is tried to be aqcuired from pool. 

Your example is also using transactions in an invalid way, that might be the reason why you are getting unhandled rejection. Check transaction provider from latest docs and set rejectioin handlers properly. 

It is not documented well, but `trx.commit()` probably never rejects (I haven't checked it from code lately though). Related issue opened about this from few weeks ago: 
https://github.com/knex/knex/issues/3583

Lets reopen if this seems to work wrong after you have fixed your transcation code.

**Comment by jakecoffman** [comment]
I updated my example in the linked repo: https://github.com/jakecoffman/knex-oracle-bug/blob/68ea525f0ea2887f50e29c1ef3a7ba0fb9b6786e/index.js

I changed it to match how you use transactions in the issue you linked. I removed the call to commit as I think that's a separate issue. 

I'm still seeing it not release the connection at the end of the script running. Here's the output when I pull the network cable during the sleep:

```
> DEBUG=* node index

  knex:tx trx1: Starting top level transaction +0ms
  knex:client acquired connection from pool: __knexUid2 +0ms
  knex:query insert into "COFFMANJ"."TABLE1" ("COLUMN1") values (?) trx1 +0ms
  knex:bindings [ 'hello' ] trx1 +0ms
Sleeping
Committing
  knex:tx trx1: releasing connection +0ms
No errors
Unhandled rejection Error: ORA-03113: end-of-file on communication channel
Process ID: 37950
Session ID: 1858 Serial number: 4295027021
```

I've tracked down what is happening:

1. It outputs [releasing connection](https://github.com/knex/knex/blob/4feefdf3d2532c23f2f662c538bf2d81700daeb1/lib/dialects/oracledb/transaction.js#L71) before it calls the driver's commit. 
1. Then it calls commit, which is going to fail because I pulled the network cable. 
1. After a bit [it is rejected](https://github.com/knex/knex/blob/4feefdf3d2532c23f2f662c538bf2d81700daeb1/lib/dialects/oracledb/index.js#L116) 
1. There is [no rejection handler in the calling code](https://github.com/knex/knex/blob/4feefdf3d2532c23f2f662c538bf2d81700daeb1/lib/dialects/oracledb/transaction.js#L73) so it is unhandled and never frees the pool. 

And actually looks like [this code](https://github.com/knex/knex/blob/4feefdf3d2532c23f2f662c538bf2d81700daeb1/lib/dialects/oracledb/transaction.js#L75) is invalid since `this` refers to the function and not the transaction instance. 

I can put a PR up for this.

**Comment by kibertoad** [comment]
Released in knex@0.20.7
